website/blog/2019-08-19-tablescan-structs.md

```diff
@@ -0,0 +1,163 @@
+
+```sql
+CREATE TABLE hive.tpch.lineitem_map (
+    l_orderkey BIGINT,
+    l_linenumber INTEGER,
+    ints map(INTEGER, BIGINT),
+    strs map(VARCHAR(2), varchar(44))
+)
+WITH (
+    format = 'ORC'
+);
+```
+
+This table has a map of 12 integers in `ints` and 5 strings in `strs`. The key is the column ordinal number in the original lineitem table as an integer in `ints` and as a string in `strs`.
+
+Like before, the tables are in ORC V2 and are compressed with Snappy. We show the Aria and baseline times as wall time seconds / CPU seconds, labeled with A: and B: respectively. 
+
+```sql
+SELECT COUNT(*), SUM(l_shipment.l_extendedprice)
+FROM exportlineitem
+WHERE l_shipment.l_partkey BETWEEN 1000000 AND 2000000 AND l_shipment.l_suppkey BETWEEN 100000 AND 200000;
+```
+A: 4 / 34
+B:  18 / 253
+
+This experiment demonstrates that a non-null struct introduces no performance penalty as compared to using top level columns. The filters are pushed down into the column scans and no struct column is materialized.
+
+``` sql
+SELECT COUNT(*), SUM(l_shipment.l_extendedprice)
+FROM exportlineitem
+WHERE l_shipment.l_partkey BETWEEN 1000000 AND 2000000 AND l_shipment.l_suppkey BETWEEN 100000 AND 200000;
+```
+A: 4 / 34
+B: 18 / 251
+
+This differs from the previous query in that we actually materialize a struct to use as input for the aggregation. This does not apreciably change the execution time since the struct `Block` only contains the extendedprice column, the rest being filled with run length encoded blocks of nulls.
+
+To further improve this we could skip making the struct completely and just project out the nested column. The difference would not be very large. We may get to this after we have done everything else.
+
+```sql
+SELECT COUNT(*), SUM(l_shipment.l_extendedprice), COUNT(l_export.license.filing_no)
+FROM exportlineitem
+WHERE l_shipment.l_suppkey BETWEEN 200000 AND 400000 AND l_shipment.l_quantity < 10;
+```
+
+A: 10 / 58
+B: 30 / 330
+
+Here we add a second struct to the mix. This takes a little longer because we must read two levels of null flags, one for l_export and the other for the license substruct. Then we read the filing number for the positions where there is a license and fill in a null for the case where either license or the enclosing l_export struct is null.
+
+# Experiments
+
+```sql
+SELECT SUM(l_shipment.l_extendedprice), COUNT(l_export.license.filing_no)
+FROM exportlineitem
+WHERE l_shipment.l_partkey BETWEEN 200000 AND 400000 AND l_shipment.l_quantity < 10 AND l_export.s_nation IN (1, 3, 6) AND l_export.c_nation = 11;
+```
+A: 5/ 50
+B: 30/309
+
+Here we add a filter on the l_export. The filters within structs are reorderable as well as the top level structs. We find that the new filter does not significantly add to the running time.
```

Comment:
This query is actually faster: 2x on wall time and 16% on CPU time. Hence, saying `new filter does not significantly add to the running time` is strange.

Reply:
Reworded to point out that it doesn't hurt and instead helps.